torch/csrc/utils/swap_thread_state.cpp

```diff
@@ -0,0 +1,23 @@
+#include <torch/csrc/utils/swap_thread_state.h>
+
+thread_local bool swappedThreadState = false;
+
+SwapPyThreadState::SwapPyThreadState() : old_tstate(PyThreadState_Get()),
+ new_tstate(PyThreadState_New(old_tstate->interp)) {
+   PyThreadState_Swap(new_tstate);
+   swappedThreadState = true;
+ }
+
+SwapPyThreadState::~SwapPyThreadState() {
+  // Copy exception to the old thread state
```

Comment:
Interesting. Is it possible to skip doing this transference, because the autograd engine is going to catch and process the Python exception in short order? (In particular, it needs to propagate the exception back to the main thread). Though I guess it probably doesn't matter too much.